File: backend/src/main/java/com/codrin/controller/ArtistsController.java
package com.codrin.controller;

import com.codrin.model.Artist;
import com.codrin.model.Response.ArtistResponse;
import com.codrin.model.Request.NewArtistRequest;
import com.codrin.model.Statistics.ArtistsSongs;
import com.codrin.service.ArtistService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/artists")
public class ArtistsController {
    private final ArtistService artistService;

    @Autowired
    public ArtistsController(ArtistService artistService) {
        this.artistService = artistService;
    }

    @GetMapping
    public ResponseEntity<List<Artist>> getArtistList(@RequestParam(required = false) Integer year) {
        if(year != null)
            return new ResponseEntity<>(artistService.getAllAfterYear(year), HttpStatus.OK);
        return new ResponseEntity<>(artistService.getAll(), HttpStatus.OK);
    }

    @GetMapping("{id}")
    public ResponseEntity<ArtistResponse> getArtist(@PathVariable("id") int id) {
        var artistResponse = artistService.getById(id);
        if(artistResponse != null)
            return new ResponseEntity<>(artistResponse, HttpStatus.OK);
        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
    }

    @PostMapping()
    public ResponseEntity<Artist> addArtist(@RequestBody NewArtistRequest request) {
        Artist artist = artistService.addArtist(request);
        return new ResponseEntity<>(artist, HttpStatus.CREATED);
    }

    @PutMapping("{id}")
    public ResponseEntity<Artist> updateArtist(@PathVariable("id") int id, @RequestBody NewArtistRequest request) {
        var artist = artistService.updateArtist(id,request);
        if(artist == null)
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        return new ResponseEntity<>(artist, HttpStatus.OK);
    }

    @DeleteMapping("{id}")
    public ResponseEntity<HttpStatus> deleteArtist(@PathVariable("id") int id) {
        var deleted = artistService.deleteArtist(id);
        if(deleted)
            return new ResponseEntity<>(HttpStatus.NO_CONTENT);
        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
    }
    @GetMapping("/songs")
    public ResponseEntity<List<ArtistsSongs>> getArtistSongs() {
        var artistSongs = artistService.getArtistsSongs();
        if(artistSongs != null)
            return new ResponseEntity<>(artistSongs, HttpStatus.OK);
        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
    }
}

File: backend/Controllers/ArtistsController.cs
using Azure;
using backend.Helpers;
using backend.Models;
using backend.Models.Extended;
using backend.Models.Request;
using backend.Models.Response;
using backend.Models.Statistics;
using backend.Services;
using backend.Wrappers;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Routing;
using Microsoft.EntityFrameworkCore;
using System.Security.Claims;

namespace backend.Controllers
{
    [ApiController]
    [Route("[controller]")]
    public class ArtistsController : Controller
    {
        private readonly ArtistsService _artistsService;
        private readonly IUriService _uriService;

        public ArtistsController(ArtistsService artistsService, IUriService uriService)
        {
            _artistsService = artistsService;
            _uriService = uriService;
        }

        [HttpGet()]
        public IActionResult GetArtistList([FromQuery] PaginationFilter filter, [FromQuery(Name = "year")] int? year)
        {
            var validFilter = new PaginationFilter(filter.PageNumber, filter.PageSize);
            List<ArtistExtended> artists = new List<ArtistExtended>();
            int totalRecords;
            if (year.HasValue)
            {
                artists = _artistsService.GetAllAfterYear(filter, year.Value);
                totalRecords = _artistsService.GetArtistsCount(year.Value);
            }
            else
            {
                artists = _artistsService.GetAll(filter);
                totalRecords = _artistsService.GetArtistsCount();
            }
            
            var route = Request.Path.Value;
            var pagedReponse = PaginationHelper.CreatePagedReponse(artists, validFilter, totalRecords, _uriService, route);

            return Ok(pagedReponse);
        }

        [HttpGet("{id}")]
        public ActionResult<ArtistResponse> GetArtist(int id)
        {
            var artist = _artistsService.GetById(id);
            if(artist == null)
            {
                return NotFound();
            }
            return Ok(artist);
        }

        [HttpPost]
        [Authorize]
        public ActionResult<Artist> PostArtist([FromBody] NewArtistRequest request)
        {
            var email = User.FindFirstValue("Email");
            Artist artist = _artistsService.AddArtist(request, email);
            return Ok(artist);
        }

        [HttpPut("{id}")]

        public ActionResult<Artist> PutArtist(int id, [FromBody] NewArtistRequest request)
        {
            if (!User.Identity.IsAuthenticated)
                return Unauthorized();
            var email = User.FindFirstValue("Email");
            var role = User.FindFirstValue(ClaimTypes.Role) == null ? "" : User.FindFirstValue(ClaimTypes.Role);
            if (!(role == "Admin" || role == "Moderator" || _artistsService.GetUserOfItem(id) == email))
                return Unauthorized();

            Artist artist = _artistsService.UpdateArtist(id, request);
            if (artist == null)
                return NotFound();
            return Ok(artist);
        }

        [HttpDelete("{id}")]
        public ActionResult DeleteArtist(int id)
        {
            if (!User.Identity.IsAuthenticated)
                return Unauthorized();
            var email = User.FindFirstValue("Email");
            var role = User.FindFirstValue(ClaimTypes.Role) == null ? "" : User.FindFirstValue(ClaimTypes.Role);
            if (!(role == "Admin" || role == "Moderator" || _artistsService.GetUserOfItem(id) == email))
                return Unauthorized();

            var deleted = _artistsService.DeleteArtist(id);
            if(deleted)
                return Ok();
            return NotFound();
        }

        [HttpPost("BulkAdd")]
        [Authorize]
        public ActionResult<List<Artist>> BulkPostArtist([FromBody] List<NewArtistRequest> requests)
        {
            List<Artist> artist = _artistsService.AddArtists(requests);
            return Ok(artist);
        }

        [HttpPost("BulkDelete")]
        [Authorize(Roles ="Admin")]
        public ActionResult<bool> BulkDeleteArtist([FromBody] List<int> ids)
        {
            bool result = _artistsService.DeleteArtists(ids);
            return Ok(result);
        }

        [HttpGet("Songs")]
        public ActionResult<List<ArtistSongs>> GetArtistsSongs([FromQuery] PaginationFilter filter)
        {
            var validFilter = new PaginationFilter(filter.PageNumber, filter.PageSize);
            List<ArtistSongs> artistsSongs = new List<ArtistSongs>();

            artistsSongs = _artistsService.GetArtistsSongs(filter);
            if (artistsSongs == null)
                return NotFound();

            var totalRecords = _artistsService.GetArtistsCount();
            var route = Request.Path.Value;
            var pagedReponse = PaginationHelper.CreatePagedReponse(artistsSongs, validFilter, totalRecords, _uriService, route);

            return Ok(pagedReponse);

        }
    }
}
